feat: Add initial packaging intro (#14)
ICESAT-2HackWeek/website-2024: book/tutorials/nb-to-package/intro.md

# Packaging notebook introduction and demo

This tutorial demonstrates how powerful packaging your Python code
can be ... even for yourself.

Before we get ahead of ourselves. Let's think about the **WHY**.
You might ask yourself:

- why would I want to package my code?
- can't I just pass around notebooks?
- what's the point of all this?
- what is the meaning of life?

... well, maybe not the last one. But the others are valid questions.

## What is a Python package?

> A "package" or officially called a "distribution package" is simply
> a directory of Python modules and sub-packages with an `__init__.py` file
> that allows for the modular organization of code.
> It can be be installed and distributed for use by others.

Reviewer: ```suggestion
A "package" is simply an organized directory of Python files that can be be installed and distributed for use by others.
```

Reviewer: (I recommend not including "module" and "sub-package" in the definition until those are defined). An alternative might be to flip the order of presentation to start with a function and build up to package

![Bedroom](https://images.unsplash.com/photo-1630699293259-0b6c08606c62?q=80&w=3870&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D)

Let's think about this in a simpler term. We can think of a
package like a room in a house. Say the "bedroom" is a package.
Inside the bedroom, you have a bed, a nightstand, and a closet.

- The bed and nightstand are like *modules* in a package.
- The closet is like a *sub-package* in a package, since there might be other
furniture pieces in a closet like a dresser; some closets can even be a whole room!

### Module

A module is a single file of Python code that can be imported into Python

> scripts or other modules,
> also known officially as an "import package".

Reviewer: ```suggestion
scripts or other modules.
```

Reviewer: (I find this really confusing as it seems like modules and libraries are "packages" so almost everything could be called a package 😅) I suggest being more opinionated for clarity about the hierarchy... Would it be too bold to claim that a "package" is strictly a "library" that can be installed with pip :) ?

Author: "modules" are not packages.. seems like that's not clear 😛 "libraries" are "packages", I've called this part out later, see https://github.com/ICESAT-2HackWeek/website-2024/pull/14/files#r1717208218. I want to highlight this fine differences hence the bedroom analogy. I'll think about how to make this more clear, but if you have any ideas, please let me know. I think it's an important difference to know.

Reviewer: I read this sentence as  `a "module" also known officially as an "import package"` I just recommend removing that last bit of the sentence

2. Then, you can import this module in another Python script or notebook:

    ```python
    from hello_world import say_hello

    say_hello()
    ```

    This will print `Hello, World!` to the console.

This is a simple example of a module, but you can have more complex modules with many functions and classes.

![Bed](https://images.unsplash.com/photo-1522771739844-6a9f6d5f14af?q=80&w=3871&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D)

Going back to the bedroom analogy, a module is like a single piece of furniture in a room.
For example, a bed is a piece of furniture that we spend most of our life on,
this "bed" might be made up of various parts like a headboard, frame, and mattress.
Each of these parts can be thought of as a function or class in a module.

### Library

A library is a collection of modules and packages that provide specific functionality or a set of related functionalities.
In the Python ecosystem, a library is often used interchangeably with the term "package", and this can be confusing.
For example, [Matplotlib](https://matplotlib.org/) is a plotting library that contains many modules to help you plot data

> or [Xarray](http://xarray.pydata.org/en/stable/) is a library that includes several modules and packages to help you work with labeled multi-dimensional arrays.

Reviewer: Another example towards the beginning of this tutorial could identify the pieces from one of these larger libraries: 
```python
from xarray.testing import assert_equal
     ^      ^              ^
   library  module         function
```
https://docs.xarray.dev/en/stable/generated/xarray.testing.assert_equal.html

It is the house itself, which contains many rooms.
Different house or even a commercial space might have different purposes and functionalities.

### The Python Terms Hierarchy

Here's a hierarchy of Python terms from smallest to largest to help you understand the relationship between modules, packages, and libraries:

<img src="../../img/code-to-package.png" alt="functions to libraries" width="200"/>

A python package may look like the following:

```console
mypackage/
    __init__.py
    module1.py
    module2.py
    subpackage/
        __init__.py
        submodule1.py
        submodule2.py

```

Within the `mypackage` package and its subpackages, you might import functions from a dependency like `numpy`
within a module file. Once this set of packages is bundled together and distributed,
you'll end up essentially with a library that can be installed and used by others.

This action of bundling your code together into a package or library is called "packaging" your code.

```{note}
For the rest of this tutorial, we'll use the term "package" to refer to a collection of code that can be installed and used by others,
which essentially means a **library**.

> ```

Author: @scottyhq This is where I called out that the rest of the tutorial I'm using "package" for library! It is confusing in the Python world.. which is unfortunate, but I'd also like to drive home that a "package" is simply a directory with an `__init__.py` file and "library" is a collection of both packages and modules.

Reviewer: All good. You could consider moving this hierarchy and note up to the top to define the terms and then jump into the analogy. I'm a visual person so I like diagrams :) But everyone's different. Happy to have you merge this as-is!

    eggsample?

    # Output
    # ...
    # String form: <module 'eggsample' ...>
    # ...
    ```
    ````

    Congratulations! You've just created a package! But we're not done yet.

4. Typically we probably have this structure where the notebook is in the same level as the package directory.
This is a common structure for a scientific project,
and totally works when you're developing a package on your local machine.

    ```console
    mynotebook.ipynb
    eggsample/
        __init__.py
        eggsellent_cook.py
    ```

    However, how can we share this package with others?

    ````{admonition} Answer
    :class: tip dropdown

    1. You might zip the `eggsample` directory and share it with others.
    2. Then they can unzip it in either:

        a. The same directory as their notebook like the structure above.

        b. A directory of their choices, say their home directory,
        then they can add the `eggsample` directory to the system path in the notebook.

        ```python
        import sys

        sys.path.insert(0, "/home/user")
        ```
    
    ````

5. At this point, you might be thinking, "This is a lot of work to share a package!"

>     And you're right! Woudn't it be nice if we could just install the package from PyPI

Author: ```suggestion
    And you're right! Wouldn't it be nice if we could just install the package from PyPI
```
